internal/handlers/common/projection.go

```diff
@@ -221,28 +221,270 @@ func projectDocumentWithoutID(doc *types.Document, projection *types.Document, i
-			time.Time, types.NullType, types.Regex, types.Timestamp: // all this types are treated as new fields value
+			time.Time, types.NullType, types.Regex, types.Timestamp: // all these types are treated as new fields value
 			projected.Set(key, value)
 
 		case bool: // field: bool
-			// process top level fields
-			if path.Len() == 1 {
-				if inclusion {
-					if docWithoutID.Has(key) {
-						projected.Set(key, must.NotFail(docWithoutID.Get(key)))
-					}
-
-					continue
+			if inclusion {
+				// inclusion projection copies the field on the path from docWithoutID to projected.
+				if _, err = includeProjection(path, docWithoutID, projected); err != nil {
+					return nil, err
 				}
 
-				projected.Remove(key)
+				continue
 			}
 
-			// TODO: process dot notation here https://github.com/FerretDB/FerretDB/issues/2430
+			// exclusion projection removes the field on the path in projected.
+			if err = excludeProjection(path, projected); err != nil {
+				return nil, err
+			}
 		default:
 			return nil, lazyerrors.Errorf("unsupported operation %s %v (%T)", key, value, value)
 		}
 	}
 
 	return projected, nil
 }
+
+// pathNotExists indicates that no path was found.
+var pathNotExists = errors.New("path not exists")
+
+// includeProjection copies the field on the path from source to projected.
+// When an array is on the path, it returns the array containing any document
+// with the same key. Dot notation with array index path does not include
+// the field unlike document.SetByPath(path).
+// Inclusion projection with non-existent path creates an empty document
+// or an empty array based on what source has.
+//
+//	Example: "v.foo" path inclusion projection:
+//	{v: {foo: 1, bar: 1}}               -> {v: {foo: 1}}
+//	{v: {bar: 1}}                       -> {v: {}}
+//	{v: [{bar: 1}]}                     -> {v: [{}]}
+//	{v: [{foo: 1}, {foo: 2}, {bar: 1}]} -> {v: [{foo: 1}, {foo: 2}, {}]}
+//
+//	Example: "v.0.foo" path inclusion projection:
+//	{v: [{foo: 1}, {foo: 2}, {bar: 1}]} -> {v: [{}, {}, {}]}
+func includeProjection(path types.Path, source any, projected *types.Document) (*types.Array, error) {
+	key := path.Prefix()
+
+	switch source := source.(type) {
+	case *types.Document:
+		embeddedSource, err := source.Get(key)
+		if err != nil {
+			// key does not exist, nothing to set.
+			return nil, nil
+		}
+
+		if path.Len() <= 1 {
+			// path reached suffix, set field in projected.
+			setBySourceOrder(key, embeddedSource, source, projected)
+			return nil, nil
+		}
+
+		doc := new(types.Document)
+
+		if projected.Has(key) {
+			// set doc if projected has field from other projection field.
+			v := must.NotFail(projected.Get(key))
+			if d, ok := v.(*types.Document); ok {
+				doc = d
+			}
+		}
+
+		// when next prefix has an arr use arr,
+		// if it has a document, doc is set by includeProjection.
+		arr, err := includeProjection(path.TrimPrefix(), embeddedSource, doc)
+
+		switch {
+		case errors.Is(err, pathNotExists):
```

Comment:
`pathNotExists` is still needed here, because we must return without setting document if path does not exist.